### PR TITLE
Making ByteBuffer's description more useful

### PR DESCRIPTION
**Pull request by supersonicbyte**
### Motivation:

Resolving the following issue: https://github.com/apple/swift-nio/issues/2863

### Modifications:

Rewrote `description` and `debugDescription` on `ByteBuffer` to make it more useful. Also fixed a bug in the in `hexDumpCompact`.

### Result:

A more useful `description` and `debugDescription`.

### CODE REVIEW

**Review by weissi — APPROVED** [review]
LGTM thank you. I'll let the NIO team actually review this before they can merge.

File: Sources/NIOCore/ByteBuffer-core.swift
Comment (weissi) [inline Sources/NIOCore/ByteBuffer-core.swift:960]:
```suggestion
    /// A `String` describing this `ByteBuffer` including length and the bytes it contains (partially).
    ///
    /// For a `ByteBuffer` initialised with `hello world`
```

Nit: The first line in the doc comments is special, it's the summary. That summary gets displayed by itself in many scenarios so we should aim for it to make sense without the rest of the doc comment.

File: Tests/NIOCoreTests/NIOAnyDebugTest.swift
Comment (weissi) [inline Tests/NIOCoreTests/NIOAnyDebugTest.swift:27]: Nit: I almost think with this change we should change NIOAny's `description` to include `ByteBuffer: ` when it's the `case .byteBuffer`. Previously that wasn't necessary because `ByteBuffer.description` contained the word `ByteBuffer` but that's now changed.

File: Tests/NIOCoreTests/ByteBufferTest.swift
Comment (weissi) [inline Tests/NIOCoreTests/ByteBufferTest.swift:3611]: think it would be nice to add two other test cases:

- the empty ByteBuffer
- one where the hex bytes get truncated (i.e. more than 128 bytes or whatever the limit is)

File: Tests/NIOCoreTests/NIOAnyDebugTest.swift
Comment (supersonicbyte) [reply Tests/NIOCoreTests/NIOAnyDebugTest.swift:27]: Looking at `NIOAny` I would improve its description altogether to include information about the wrapped type. But I would do that in a separate issue/PR.

File: Tests/NIOCoreTests/ByteBufferTest.swift
Comment (supersonicbyte) [reply Tests/NIOCoreTests/ByteBufferTest.swift:3611]: Thanks, added!

**Review by FranzBusch — APPROVED** [review]
This LGTM. @supersonicbyte please raise another PR for adapting `NIOAny` then

**Pull request merged**
